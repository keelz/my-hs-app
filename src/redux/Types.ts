import { MiddlewareAPI, Dispatch } from 'redux';
import { IBrowser } from 'redux-responsive';
import { AppState } from './reducers/App';
import { ICollectionState } from './reducers/Collection';
import { ICardsState } from './reducers/Cards';

export interface IActionType<T> {
  type: string;
  payload?: T;
}

export interface IRootState {
  App: AppState;
  Browser: IBrowser;
  Cards: ICardsState;
  Collection: ICollectionState;
}

export type AppMiddlewareApi = MiddlewareAPI<Dispatch<IActionType<any>>, IRootState>;

export type AsyncMiddlewareOperation = (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => Promise<any>;

export type MiddlewareOperation = (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => void;

export type SafeMiddlewareOperation = (api: AppMiddlewareApi) =>
  (action: IActionType<any>) => void;

export type CollectionFilters = {
  [field: string]: string | string[]
};

export enum MODAL_STATE { CLOSED, OPEN }

export enum PLAY_STYLE {
  STANDARD = 'STANDARD',
  WILD = 'WILD',
}
