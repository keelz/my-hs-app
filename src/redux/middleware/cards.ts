import { Dispatch } from 'redux';
import { AppMiddlewareApi, IActionType } from '../Types';
import cardsService, { ICardsService } from '../../common/services/cards.service';
import {
  CARDS_SET_ACTIVE_CLASSNAME,
  CARDS_SET_CARDS,
  CARDS_SET_FILTER,
  CARDS_SET_PAGINATION,
} from '../reducers/Cards';

export const composeMiddleware = (service: ICardsService) =>
  (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => {
    switch (action.type) {
      case CARDS_SET_ACTIVE_CLASSNAME:
        return service.handleSetActiveClassname(api)(next)(action);
      case CARDS_SET_CARDS:
        next(action);
        service.resetPagination(api)(action);
        return;
      case CARDS_SET_FILTER:
        return service.handleSetFilter(api)(next)(action);
      case CARDS_SET_PAGINATION:
        return service.handleSetPagination(api)(next)(action);
      default:
        return next(action);
    }
  };

const middleware = composeMiddleware(cardsService);

export default middleware;
