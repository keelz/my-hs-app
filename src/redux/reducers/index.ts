import { combineReducers } from 'redux';
import { responsiveStateReducer as Browser } from 'redux-responsive';
import appReducer from './App';

const rootReducer = combineReducers({
  app: appReducer,
  browser: Browser,
});

export default rootReducer;
