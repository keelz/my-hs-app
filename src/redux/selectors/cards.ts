import { createSelector } from 'reselect';
import HSJSON from '../../common/constants/hsJson';
import { IRootState } from '../Types';
import { ICard } from '../../common/models/Card';
import { composeCardClassNames } from '../../common/utils/card';

export class Accessors {
  /**
   * get active card class name.
   * @param state {IRootState}
   */
  public static getActiveCardClassName(state: IRootState): string {
    return state.Cards.activeClassName;
  }

  /**
   * get all cards.
   * @param state {IRootState}
   */
  public static getCards(state: IRootState): ICard[] {
    return state.Cards.data;
  }

  /**
   * get all Cards filters.
   * @param state {IRootState}
   */
  public static getFilters(state: IRootState): { [key: string]: string } {
    return state.Cards.filters;
  }
}

/**
 * get active card class name.
 * @param state {IRootState}
 */
export const selectActiveCardClassName = (state: IRootState): string =>
  createSelector(
    [Accessors.getActiveCardClassName],
    (className) => className
  )(state);

/**
 * get all unique card class names.
 * @param state {IRootState}
 */
export const selectCardClassNames = (state: IRootState) =>
  createSelector(
    [Accessors.getCards],
    (cards) => composeCardClassNames(cards)
  )(state);

/**
 * get all cards.
 * @param state {IRootState}
 */
export const selectCards = (state: IRootState): ICard[] =>
  createSelector(
    [Accessors.getCards],
    (cards) => cards
  )(state);

/**
 * get cards for active class name.
 * @param state {IRootState}
 */
export const selectCardsWithFilters = (state: IRootState): ICard[] =>
  createSelector(
    [
      Accessors.getActiveCardClassName,
      Accessors.getCards,
      Accessors.getFilters,
    ],
    (
      activeClassName: string,
      cards: ICard[],
      filters: { [key: string]: string }
    ) => cards
      // filter by active class name
      .filter(card => card.cardClass === activeClassName)
      // filter by card cost
      .filter(card => {
        const filter = filters[HSJSON.RESPONSE_PARAMS.COST];
        const response = !!filter
          ? parseInt(filter, 10) === 7
            ? card.cost >= 7
            : card.cost === parseInt(filter, 10)
          : true;
        return response;
      })
      // sort by cost
      .sort((a, b) => a.cost - b.cost)
      // sort by name
      .sort((a, b) => a.cost === b.cost
        ? a.name < b.name
          ? -1
          : 0
        : 0)
  )(state);
