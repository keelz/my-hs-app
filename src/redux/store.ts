import { createBrowserHistory } from 'history';
import { routerMiddleware } from 'react-router-redux';
import { applyMiddleware, createStore, compose } from 'redux';
import { responsiveStoreEnhancer } from 'redux-responsive';
import createDebounce from 'redux-debounced';
import thunk from 'redux-thunk';
import appMiddleware from './middleware/App';
import rootReducer from './reducers';

/**
 * redux devtools implementation.
 * We want to compose our enhancers normally within the production
 * environment (build), otherwise we want to compose our ehancers
 * with the devtools extension.
 */
export const composeFunc = () => {
  return process.env.NODE_ENV === 'production'
    ? compose
    : (() => {
      const { composeWithDevTools } = require('redux-devtools-extension');
      return composeWithDevTools;
    })();
};

/**
 * Third party redux middleware
 * - redux-thunk
 * - router
 * - react-router-redux
 * - history
 * - redux-responsive
 */
const libEnhancers = [
  thunk,
  routerMiddleware(createBrowserHistory({ basename: '/' })),
  createDebounce(),
].map(mw => applyMiddleware(mw));
libEnhancers.push(responsiveStoreEnhancer);

/**
 * Custom middleware
 * - appMiddleware
 */
const customEnhancers = [
  appMiddleware,
].map(mw => applyMiddleware(mw));

// compose enhancers with composeFunc
// https://redux.js.org/api/compose
const composedEnhancers = composeFunc().apply(null, [...libEnhancers, ...customEnhancers]);

// create store
// https://redux.js.org/api/createstore
const store = createStore(rootReducer, {}, composedEnhancers);

export default store;
