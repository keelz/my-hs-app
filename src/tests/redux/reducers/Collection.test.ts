import reducer, {
  collectionStateFactory,
  setCollectionActiveCardAction,
  setCollectionModalAction,
  COLLECTION_SET_ACTIVE_CARD,
  COLLECTION_SET_MODAL,
} from '../../../redux/reducers/Collection';
import { ModalState } from '../../../redux/Types';
import { CardClassName } from '../../../common/models/cards.model';

const testCard = {
  cardClass: CardClassName.DRUID,
  cost: 4,
  flavor: 'test',
  id: 'test',
  name: 'test',
  text: 'test',
};

describe('Collection', () => {
  describe('factories', () => {
    it('composes state with factory correctly', () => {
      const state = collectionStateFactory();
      expect(state).toMatchSnapshot();
    });
  });

  describe('action creators', () => {
    it(`dispatches ${COLLECTION_SET_ACTIVE_CARD} correctly`, () => {
      const dispatch = jest.fn();
      setCollectionActiveCardAction(testCard)(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: COLLECTION_SET_ACTIVE_CARD,
        payload: { activeCard: { ...testCard } },
      });
    });

    it(`dispatches ${COLLECTION_SET_MODAL} correctly`, () => {
      const dispatch = jest.fn();
      setCollectionModalAction(ModalState.OPEN)(dispatch);
      expect(dispatch).toHaveBeenCalledWith({
        type: COLLECTION_SET_MODAL,
        payload: { modal: ModalState.OPEN },
      });
    });
  });

  describe('reducer', () => {
    it('composes default state correctly', () => {
      const action = {
        type: 'UNKNOWN',
      };
      const state = reducer(undefined, action);
      expect(state).toEqual(collectionStateFactory());
    });

    it(`composes state for ${COLLECTION_SET_ACTIVE_CARD} correctly`, () => {
      const action = {
        type: COLLECTION_SET_ACTIVE_CARD,
        payload: {
          ...collectionStateFactory(),
          activeCard: testCard,
        },
      };
      const state = reducer(undefined, action);
      expect(state).toEqual(action.payload);
    });

    it(`composes state for ${COLLECTION_SET_MODAL} correctly`, () => {
      const action = {
        type: COLLECTION_SET_MODAL,
        payload: { modal: ModalState.OPEN },
      };
      const state = reducer(collectionStateFactory(), action);
      expect(state).toEqual({
        ...collectionStateFactory(),
        modal: ModalState.OPEN,
      });
    });
  });
});
