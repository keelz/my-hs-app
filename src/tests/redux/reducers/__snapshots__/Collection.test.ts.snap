// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection factories composes state with factory correctly 1`] = `
Object {
  "activeCard": undefined,
  "activeClassName": "DRUID",
  "filters": Object {
    "PLAY_STYLE": "STANDARD",
  },
  "modal": 0,
  "pagination": Object {
    "currentPage": 0,
    "itemsPerPage": 0,
    "pages": 0,
    "total": 0,
  },
}
`;
